File: Codeforces Round #856 (Div. 2)/D.cpp
// https://codeforces.com/contest/1794/problem/D

#include <bits/stdc++.h>

using namespace std;

typedef unsigned long long int ulong;

static inline pair<ulong, vector<ulong>> parse(void) {
  ulong n;
  cin >> n;

  vector<ulong> as(2 * n);
  for (ulong i = 0; i < 2 * n; ++i) {
    cin >> as[i];
  }

  return {n, as};
}

#define mod 998244353

static inline bool* primes_until(ulong N) {
  bool* primes = (bool*)(calloc(N, sizeof(bool)));

  memset(primes, true, N);

  primes[0] = false;
  primes[1] = false;

  for (ulong i = 2; i * i < N; i++) {
    if (primes[i]) {
      for (ulong j = i * i; j < N; j += i) {
        primes[j] = false;
      }
    }
  }

  return primes;
}

static inline ulong factorial_mod(ulong x) {
  ulong result = 1;
  for (ulong i = 2; i <= x; ++i) {
    result = (result * i) % mod;
  }
  return result;
}

static inline ulong pow_mod(ulong x, ulong y) {
  ulong res = 1;
  while (y > 0) {
    if (y % 2 == 1) {
      res = (res * x) % mod;
    }
    y = y / 2;
    x = (x * x) % mod;
  }
  return res;
}

static inline ulong inverse_mod(ulong x) {
  return pow_mod(x, mod - 2);
}

static inline ulong div_mod(ulong x, ulong y) {
  return ((x % mod) * (inverse_mod(y % mod))) % mod;
}

static inline ulong prod_mod(vector<ulong> xs) {
  ulong result = 1;
  for (ulong x : xs) {
    result = (result * x) % mod;
  }
  return result;
}

static inline ulong prod_facts_values_mod(map<ulong, ulong> xs) {
  ulong result = 1;
  for (pair<const ulong, ulong> x : xs) {
    result = (result * x.second) % mod;
  }
  return result;
}

ulong as_primes_combinations[4046][2023];

static inline ulong solve(ulong n, vector<ulong> as) {

  bool* primes = primes_until(1000005);

  map<ulong, ulong> as_primes = {};
  map<ulong, ulong> as_non_primes = {};

  for (ulong a : as) {
    if (primes[a]) {
      as_primes[a] += 1;
    } else {
      as_non_primes[a] += 1;
    }
  }

  ulong amount_primes = as_primes.size();
  if (amount_primes < n) {
    return 0;
  }

  ulong factorial_mod_n = factorial_mod(n);

  vector<ulong> as_primes_amounts = {};
  for (pair<const ulong, ulong> x : as_primes) {
    as_primes_amounts.push_back(x.second);
  }
  ulong prod_facts_amounts_non_primes = prod_facts_values_mod(as_non_primes);

  as_primes_combinations[0][0] =
      div_mod(factorial_mod_n, prod_facts_amounts_non_primes);
  for (ulong available_primes = 1; available_primes <= amount_primes;
       ++available_primes) {
    as_primes_combinations[available_primes][0] = div_mod(
        factorial_mod_n,
        prod_facts_amounts_non_primes *
            factorial_mod(as_primes_amounts[available_primes - 1])
    );
  }

  for (ulong needed_primes = 1; needed_primes <= n; needed_primes++) {
    as_primes_combinations[0][needed_primes] = 0;
  }

  for (ulong available_primes = 1; available_primes <= amount_primes;
       ++available_primes) {
    for (ulong needed_primes = 1; needed_primes <= min(n, available_primes);
         ++needed_primes) {
      as_primes_combinations[available_primes][needed_primes] =
          (div_mod(
               as_primes_combinations[available_primes - 1][needed_primes - 1],
               factorial_mod(as_primes_amounts[available_primes - 1] - 1)
           ) +
           div_mod(
               as_primes_combinations[available_primes - 1][needed_primes],
               factorial_mod(as_primes_amounts[available_primes - 1])
           )) %
          mod;
    }
  }

  return as_primes_combinations[amount_primes][n];
}

int main(void) {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);

  pair<ulong, vector<ulong>> input = parse();
  ulong res = solve(input.first, input.second);
  cout << res << endl;

  return EXIT_SUCCESS;
}

File: Codeforces Round #856 (Div. 2)/D.py
# https://codeforces.com/contest/1794/problem/D

from math import factorial, prod
from typing import Dict, List, Tuple

mod: int = 998244353

def parse() -> Tuple[int, List[int]]:
    n = int(input())
    return n, list(map(int, input().split()))

def primes_until(n: int) -> List[bool]:
    res: List[bool] = [True] * (n + 1)
    res[0] = False
    res[1] = False
    for i in range(2, n + 1):
        if res[i]:
            for j in range(i * i, n + 1, i):
                res[j] = False

    return res

def factorial_mod(n: int) -> int:
    res: int = 1
    for i in range(2, n + 1):
        res = (res * i) % mod
    return res

def pow_mod(x: int, y: int) -> int:
    res: int = 1
    while y > 0:
        if y % 2 == 1:
            res = (res * x) % mod
        y = y // 2
        x = (x * x) % mod
    return res

def inverse_mod(x: int) -> int:
    return pow_mod(x, mod - 2)

def div_mod(x: int, y: int) -> int:
    res = ((x % mod) * (inverse_mod(y % mod))) % mod
    return res

def prod_mod(it: List[int]) -> int:
    res: int = 1
    for i in it:
        res = (res * i) % mod
    return res

def solve(n: int, As: List[int]) -> int:
    primes: List[bool] = primes_until(10 ** 6)

    As_primes: Dict[int, int] = {}
    As_non_primes: Dict[int, int] = {}
    for a in As:
        if primes[a]:
            if a in As_primes.keys():
                As_primes[a] += 1
            else:
                As_primes[a] = 1
        else:
            if a in As_non_primes.keys():
                As_non_primes[a] += 1
            else:
                As_non_primes[a] = 1

    amount_primes: int = len(As_primes)

    if amount_primes < n:
        return 0

    As_primes_combinations: List[List[float]] = [[0] * (n + 1) for _ in range(amount_primes + 1)]

    factorial_mod_n: int = factorial_mod(n)
    as_primes_amounts: List[int] = list(As_primes.values())
    prod_facts_amounts_non_primes: int = prod_mod([factorial_mod(v) for v in As_non_primes.values()])

    As_primes_combinations[0][0] = div_mod(factorial_mod_n, prod_facts_amounts_non_primes)
    for available_primes in range(1, amount_primes + 1):
        As_primes_combinations[available_primes][0] = div_mod(
            factorial_mod_n,
            prod_facts_amounts_non_primes * factorial_mod(as_primes_amounts[available_primes - 1])
        )

    for needed_primes in range(1, n + 1):
        As_primes_combinations[0][needed_primes] = 0

    for available_primes in range(1, amount_primes + 1):
        for needed_primes in range(1, min(n + 1, available_primes + 1)):
            As_primes_combinations[available_primes][needed_primes] = (
                (div_mod(
                    As_primes_combinations[available_primes - 1][needed_primes - 1],
                    factorial_mod(list(As_primes.values())[available_primes - 1] - 1)
                ) + div_mod(
                    As_primes_combinations[available_primes - 1][needed_primes],
                    factorial_mod(list(As_primes.values())[available_primes - 1])
                )) % mod
            )

    res = As_primes_combinations[amount_primes][n]

    return res



def main():
    print(solve(*parse()))


if __name__ == '__main__':
    main()
